port/unix/omrsysinfo.c

```diff
@@ -5416,6 +5417,27 @@ isCgroupV1Available(struct OMRPortLibrary *portLibrary)
 	return result;
 }
 
+/**
+ * @internal
+ * Checks if cgroup v2 system is available
+ *
+ * @param[in] portLibrary pointer to OMRPortLibrary
+ *
+ * @return TRUE if cgroup v2 system is available, FALSE otherwise
+ */
+static BOOLEAN
+isCgroupV2Available(struct OMRPortLibrary *portLibrary)
```

Comment:
Would a static int make sense here? `static int available = -1;` Set to 0 or 1 on the first pass so we can check for -1

Reply:
Follow the example of `PPG_cgroupEntryList` to add a new global to cache these values.

I will go with a bit map:
```
OMR_CGROUP_ISAVAILABLE_V1_SET 0x1
OMR_CGROUP_ISAVAILABLE_V2_SET 0x2
OMR_CGROUP_ISAVAILABLE_V1 0x4
OMR_CGROUP_ISAVAILABLE_V2 0x8

PPG_cgroupControlFlags

isCgroupV1Available: ALL_BITS_SET(PPG_cgroupControlFlags, OMR_CGROUP_ISAVAILABLE_V1 | OMR_CGROUP_ISAVAILABLE_V1_SET)

isCgroupV2Available: ...

# Mutually exclusive
Assert(ALL_BITS_SET(PPG_cgroupControlFlags, OMR_CGROUP_ISAVAILABLE_V1 | OMR_CGROUP_ISAVAILABLE_V2)
```

We will need to account for synchronization/locking since this is a global and OMR cgroup functions can be invoked concurrently by multiple threads. `PPG_cgroupEntryList` uses a `omrthread_monitor_t` which is heavy duty. The light-weight alternative is the use of atomic ops: [VM_AtomicSupport::get/set](https://github.com/eclipse/omr/blob/def6c5a67a3637fa49342227ca9d69591be7c261/include_core/AtomicSupport.hpp#L920).

The other alternative is to initialize the bitmap at startup: `omrsysinfo_startup`. Follow the `isRunningInContainer` example. With this approach, we won't need synchronization since the OMR cgroup functions won't be invoked at startup. The bitmap will always be initialized whenever the OMR cgroup functions are invoked; so, we won't need the `_SET` states in the bitmap. I will go with the startup approach because I believe it's inexpensive to evaluate `isCgroupV1Available` and `isCgroupV2Available`, and there will be unnoticeable impact on startup.

On another note, `omrsysinfo_cgroup_is_system_available` invokes `isRunningInContainer`. Since `omrsysinfo_startup` caches its value in `PPG_isRunningInContainer`, we can reuse the cached value in `omrsysinfo_cgroup_is_system_available` instead of reinvoking `isRunningInContainer`.

We can also represent `PPG_isRunningInContainer` as a bit in `PPG_cgroupControlFlags (uintptr_t)`:
```
OMR_CGROUP_IS_V1_AVAILABLE 0x1
OMR_CGROUP_IS_V2_AVAILABLE 0x2
OMR_IS_RUNNING_IN_CONTAINER 0x4
...
```

Reply:
PPG_isRunningInContainer is not necessarily related to cgroups. It is defined in a different spot compared to e.g. PPG_cgroupEntryList and also not ifdef'd with `LINUX`. I think it would be best to leave it as is.

Reply:
On second thought, should we replace the `isRunningInContainer` call in `omrsysinfo_cgroup_is_system_available` with the bit check? Currently if the `isRunningInContainer` call in `omrsysinfo_cgroup_is_system_available` fails, the function is aborted but if we replace it with a bit check, `populateCgroupEntryListV1` will be called even if `isRunningInContainer` fails.

Reply:
`isRunningInContainer` should always work. If it fails, then there is either something wrong in the environment or it needs to be corrected. Failing at startup or `omrsysinfo_cgroup_is_system_available` should be irrelevant. Also, the error gets ignored when `isRunningInContainer` fails in `omrsysinfo_cgroup_is_system_available` (just false is returned). An error should not be ignored. It should be caught and fixed as soon as possible before customers are impacted.

Reply:
We should be careful about introducing accesses to the filesystem unconditionally. See the discussion in openjdk/jdk#7825 which noted that reading `/proc/cpuinfo` sometimes takes a significant amount of time.

Reply:
https://github.com/openjdk/jdk/pull/7825 mentions of regression via `fopen`. `isCgroupV1Available` uses `statfs` and `isCgroupV2Available` uses `access`. `statfs` and `access` are almost equivalent and cheaper than `fopen`. But, `isRunningInContainer` relies upon `fopen`, and it is invoked from `omrsysinfo_startup`. @EricYangIBM Can you do perf comparisons?
```
JVM-1: invokes isCgroupV1Available, isCgroupV2Available and isRunningInContainer from omrsysinfo_startup
JVM-2: disables the above calls in omrsysinfo_startup

time JVM-1/bin/java -version
time JVM-2/bin/java -version
```
Since we are only evaluating the startup cost of the above functions, the full cgroup API does not need to function correctly. We just need to make sure the functions, whose cost is being evaluated, are invoked at startup.

Reply:
Latest commit adds the global

Reply:
On my fyre machine, there is no difference in perf, but inside a container in the same machine, time with the 3 functions is ~0.150s  whereas time without the functions is ~0.130s. Also, the docker times have a larger range (+/-0.05 vs +/- 0.02). Only including the `isCgroupV1Available` functions doesn't affect the time. I've narrowed it down to `isRunningInContainer`'s `access`. This function is already in the master branch, have we done any perf runs that could expose this possible regression?

Reply:
Perf work is mostly done on bare metal machines. Fyre machines are virtualized. If no perf regression is seen on a Fyre machine, then a perf regression won't be seen in our perf builds which run on bare metal machines.

On a Fyre machine, starting a container will add a second layer of virtualization. The extra layer of virtualization makes `isRunningInContainer`'s cost very noticeable.

Since `isRunningInContainer` was already invoked during startup and your new `isCgroupV*Available` functions do not influence perf, the startup perf issue can be addressed in a separate issue/PR. But, we should document it in #1281. @keithc-ca do you agree?